src/ripple/app/tx/impl/AMMDeposit.cpp

```diff
@@ -0,0 +1,528 @@
+    auto const lpTokens = ctx.tx[~sfLPToken];
+    // Valid options are:
+    //   LPTokens
+    //   Asset1In
+    //   Asset1In and Asset2In
+    //   Asset1In and LPTokens
+    //   Asset1In and EPrice
+    if ((asset1In && asset2In && (lpTokens || ePrice)) ||
+        (asset1In && lpTokens && (asset2In || ePrice)) ||
+        (asset1In && ePrice && (asset2In || lpTokens)) ||
+        (ePrice && !asset1In) || (asset2In && !asset1In) ||
+        (!lpTokens && !asset1In))
+    {
+        JLOG(ctx.j.debug()) << "AMM Deposit: invalid combination of "
+                               "deposit fields.";
+        return temBAD_AMM_OPTIONS;
+    }
+
+    if (lpTokens && *lpTokens <= beast::zero)
+    {
+        JLOG(ctx.j.debug()) << "AMM Deposit: invalid LPTokens";
+        return temBAD_AMM_TOKENS;
+    }
+
+    if (auto const res = invalidAmount(asset1In, (lpTokens || ePrice)))
+    {
+        JLOG(ctx.j.debug()) << "AMM Deposit: invalid Asset1In";
+        return *res;
+    }
+
+    if (auto const res = invalidAmount(asset2In))
+    {
+        JLOG(ctx.j.debug()) << "AMM Deposit: invalid Asset2InAmount";
+        return *res;
+    }
+
+    if (auto const res = invalidAmount(ePrice))
+    {
+        JLOG(ctx.j.debug()) << "AMM Deposit: invalid EPrice";
+        return *res;
+    }
+
+    return preflight2(ctx);
+}
+
+TER
+AMMDeposit::preclaim(PreclaimContext const& ctx)
+{
+    auto const accountID = ctx.tx[sfAccount];
+
+    if (!ctx.view.read(keylet::account(accountID)))
+    {
+        JLOG(ctx.j.debug()) << "AMM Deposit: Invalid account.";
+        return terNO_ACCOUNT;
+    }
+
+    auto const ammSle = getAMMSle(ctx.view, ctx.tx[sfAMMID]);
+    if (!ammSle)
+    {
+        JLOG(ctx.j.debug()) << "AMM Deposit: Invalid AMMID.";
+        return terNO_ACCOUNT;
+    }
+
+    auto const asset1In = ctx.tx[~sfAsset1In];
+    auto const asset2Out = ctx.tx[~sfAsset2Out];
+
+    if (asset1In)
+    {
+        if (auto const ter =
+                requireAuth(ctx.view, asset1In->issue(), accountID);
+            ter != tesSUCCESS)
+        {
+            JLOG(ctx.j.debug()) << "AMM Deposit: account is not authorized, "
+                                << asset1In->issue();
+            return ter;
+        }
+    }
+
+    if (asset2Out)
+    {
+        if (auto const ter =
+                requireAuth(ctx.view, asset2Out->issue(), accountID);
+            ter != tesSUCCESS)
+        {
+            JLOG(ctx.j.debug()) << "AMM Deposit: account is not authorized, "
+                                << asset2Out->issue();
+            return ter;
+        }
+    }
+
+    if (isFrozen(ctx.view, asset1In) || isFrozen(ctx.view, asset2Out))
+    {
+        JLOG(ctx.j.debug()) << "AMM Deposit involves frozen asset.";
+        return tecFROZEN;
+    }
+
+    auto const [asset1, asset2, lptAMMBalance] =
+        ammHolds(ctx.view, *ammSle, std::nullopt, std::nullopt, ctx.j);
+    if (asset1 <= beast::zero || asset2 <= beast::zero ||
```

Comment:
Do we need to check that the assets on the transaction match the assets in the amm?